assets/app/scripts/services/apiService.js

```diff
@@ -0,0 +1,268 @@
+      }
+      return $window.location.protocol === "http:" ? "http" : "https";
+    };
+
+
+    var cleanCopyParams = function(params) {
+      params = (params &&  angular.copy(params)) || {};
+      delete params.namespace;
+      return params;
+    };
+
+
+    // generates something like:
+    // https://localhost:8443
+    var openshiftAPIBaseUrl = function() {
+      var protocol = $window.location.protocol === "http:" ? "http" : "https";
+      var hostPort = API_CFG.openshift.hostPort;
+      return new URI({protocol: protocol, hostname: hostPort}).toString();
+    };
+
+    // TODO: can this be implmeneted in a meangful way at this point?
+    var resourceInfo = function(resource) {
+      console.warn('TODO: resourceInfo is not implemented....!', !!findAPIFor(qualifyResource(resource)));
+      return !!findAPIFor(qualifyResource(resource));
+    };
+
+
+    // port of restmapper.go#normalizeResource
+    // TODO: upate to match the current go version func of the same name
+    var normalizeResource = function(resource) {
+       if (!resource) {
+        return;
+       }
+       // only lowercase the first segment, leaving subresources as-is (some are case-sensitive)
+       var segments = resource.split("/");
+       segments[0] = segments[0].toLowerCase();
+       var normalized = segments.join("/");
+       if (resource !== normalized) {
+         Logger.warn('Non-lower case resource "' + resource + '"');
+       }
+       return normalized;
+    };
+
+
+    // port of restmapper.go#kindToResource,
+    // TODO: upate to match the current go version func of the same name
+    var kindToResource = function(kind) {
+      if (!kind) {
+        return "";
+      }
+      var resource = String(kind).toLowerCase();
+
+      if (_.endsWith(resource, 'status')) {
+        resource = resource + 'es';
+      } else if (_.endsWith(resource, 's')) {
+        // no-op
+      } else if (_.endsWith(resource, 'y')) {
+        resource = resource.substring(0, resource.length-1) + 'ies';
+      } else {
+        resource = resource + 's';
+      }
+      // NOTE: previously had a check here for a 'known resource'
+      return resource;
+    };
+
+
+    // if resource is string will convert to the newer object structure {resource: '', group: '', version: ''}
+    // TODO: ensure kindToResouce() has been handed already...
+    var qualifyResource = function(resourceWithSubresource, apiVersion) {
+
+      if(!resourceWithSubresource) {
+        return;
+      }
+      var qualified = _.isString(resourceWithSubresource) ? { resource: resourceWithSubresource } : _.clone(resourceWithSubresource); // clone for manipulation
+
+      //qualified.resource = kindToResource(qualified.kind || resource);
+      //qualified.resource = normalizeResource((qualified.kind && kindToResource(qualified.kind)) || qualified.resource);
+      qualified.resource = normalizeResource(qualified.kind || qualified.resource);
+
+      if(!qualified.group) {
```

Comment:
and only do that if `qualified.group` and `qualified version` aren't already set